src/appleseed-max-impl/appleseedvolumemtl/appleseedvolumemtl.cpp

```diff
@@ -0,0 +1,593 @@
+{
+    return nullptr;
+}
+
+void AppleseedVolumeMtl::SetSubTexmap(int i, Texmap* texmap)
+{
+}
+
+int AppleseedVolumeMtl::MapSlotType(int i)
+{
+    return MAPSLOT_TEXTURE;
+}
+
+MSTR AppleseedVolumeMtl::GetSubTexmapSlotName(int i)
+{
+    return L"";
+}
+
+void AppleseedVolumeMtl::Update(TimeValue t, Interval& valid)
+{
+    if (!m_params_validity.InInterval(t))
+    {
+        m_params_validity.SetInfinite();
+
+        NotifyDependents(FOREVER, PART_ALL, REFMSG_CHANGE);
+    }
+
+    valid &= m_params_validity;
+}
+
+void AppleseedVolumeMtl::Reset()
+{
+    g_appleseed_volumemtl_classdesc.Reset(this);
+
+    m_params_validity.SetEmpty();
+}
+
+Interval AppleseedVolumeMtl::Validity(TimeValue t)
+{
+    Interval valid = FOREVER;
+    Update(t, valid);
+    return valid;
+}
+
+ParamDlg* AppleseedVolumeMtl::CreateParamDlg(HWND hwMtlEdit, IMtlParams* imp)
+{
+    ParamDlg* param_dialog = g_appleseed_volumemtl_classdesc.CreateParamDlgs(hwMtlEdit, imp, this);
+    DbgAssert(m_pblock != nullptr);
+    return param_dialog;
+}
+
+IOResult AppleseedVolumeMtl::Save(ISave* isave)
+{
+    bool success = true;
+
+    isave->BeginChunk(ChunkFileFormatVersion);
+    success &= write(isave, FileFormatVersion);
+    isave->EndChunk();
+
+    isave->BeginChunk(ChunkMtlBase);
+    success &= MtlBase::Save(isave) == IO_OK;
+    isave->EndChunk();
+
+    return success ? IO_OK : IO_ERROR;
+}
+
+IOResult AppleseedVolumeMtl::Load(ILoad* iload)
+{
+    IOResult result = IO_OK;
+
+    while (true)
+    {
+        result = iload->OpenChunk();
+        if (result == IO_END)
+            return IO_OK;
+        if (result != IO_OK)
+            break;
+
+        switch (iload->CurChunkID())
+        {
+          case ChunkFileFormatVersion:
+            {
+                USHORT version;
+                result = read<USHORT>(iload, &version);
+            }
+            break;
+
+          case ChunkMtlBase:
+            result = MtlBase::Load(iload);
+            break;
+        }
+
+        if (result != IO_OK)
+            break;
+
+        result = iload->CloseChunk();
+        if (result != IO_OK)
+            break;
+    }
+
+    return result;
+}
+
+Color AppleseedVolumeMtl::GetAmbient(int mtlNum, BOOL backFace)
+{
+    return Color(0.0f, 0.0f, 0.0f);
+}
+
+Color AppleseedVolumeMtl::GetDiffuse(int mtlNum, BOOL backFace)
+{
+    Color color;
+    m_pblock->GetValue(ParamIdVolumeScattering, GetCOREInterface()->GetTime(), color, FOREVER);
+    return color;
+}
+
+Color AppleseedVolumeMtl::GetSpecular(int mtlNum, BOOL backFace)
+{
+    return Color(0.0f, 0.0f, 0.0f);
+}
+
+float AppleseedVolumeMtl::GetShininess(int mtlNum, BOOL backFace)
+{
+    return 0.0f;
+}
+
+float AppleseedVolumeMtl::GetShinStr(int mtlNum, BOOL backFace)
+{
+    return 0.0f;
+}
+
+float AppleseedVolumeMtl::GetXParency(int mtlNum, BOOL backFace)
+{
+    float multiplier;
+    m_pblock->GetValue(ParamIdVolumeScatteringMultiplier, GetCOREInterface()->GetTime(), multiplier, FOREVER);
```

Comment:
We can do better by taking into using the luminance of `scattering color * scattering multiplier` instead of just relying on the multiplier.